### PR TITLE
[Feature] Support engine with NPU backend.

### PR DESCRIPTION
**Pull request by wangjiangben-hw**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Added ascending device support in mmcv.

## Modification

- The main modification points are as follows: 
    1. In ```mmengine/device/utils.py```, We've added our NPU devices, When judging whether the device is available, try to import the npu plugin package.
    2. In ```mmengine/dist/```, We have added NPU-related distributed initialization methods.
    3. In ```mmengine/optim/optimizer/``` and ```mmengine/runner/amp.py```, We've added some judgment to make the mixed precision mode turn on automatically on the NPU and work properly.

## BC-breaking (Optional)

None

## Use cases (Optional)

- test case
    - tests/test_device/test_device.py
    
## Checklist

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @wangjiangben-hw 

Thanks for your kind contribution. Would you like to fix the lint issue so that we could merge this PR?
As this issue is fixed in the most recent main branch, you cam simply fix it by rebasing your code onto the most recent main.

File: mmengine/dist/dist.py [hunk 1]
```diff
@@ -411,7 +412,10 @@ def _broadcast_object_list(object_list: List[Any],
     group_backend = get_backend(group)
     is_nccl_backend = group_backend == torch_dist.Backend.NCCL
     current_device = torch.device('cpu')
-    if is_nccl_backend:
+    if is_npu_available():
+        current_device = torch.npu.current_device()
+        object_sizes_tensor = object_sizes_tensor.to(current_device)
+    elif is_nccl_backend:
```
Comment (zhouzaida) [inline mmengine/dist/dist.py:418]: Suggest making `get_backend` also support npu backend.

https://github.com/open-mmlab/mmengine/blob/13484aae9fbe66dd98f6a645267d9965239b7ce8/mmengine/dist/utils.py#L176

And then we can make this block consistent.

```python
    group_backend = get_backend(group)
    is_nccl_backend = group_backend == torch_dist.Backend
    is_hccl_backend = group_backend == 'hccl'
    if is_hccl_backend :
        current_device = torch.npu.current_device()
        object_sizes_tensor = object_sizes_tensor.to(current_device)
    elif is_nccl_backend:
```

File: mmengine/dist/dist.py [hunk 2]
```diff
@@ -430,7 +434,7 @@ def _broadcast_object_list(object_list: List[Any],
             dtype=torch.uint8,
         )
 
-    if is_nccl_backend:
+    if is_nccl_backend or is_npu_available():
```
Comment (zhouzaida) [inline mmengine/dist/dist.py:437]:
```suggestion
    if is_nccl_backend or is_hccl_backend:
```

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 3]
```diff
@@ -3,13 +3,18 @@
 
 import torch
 import torch.nn as nn
-from torch.cuda.amp import GradScaler
 
+from mmengine.device import is_npu_available
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:7]:
```suggestion
from mmengine.device import is_npu_available, is_cuda_available
```

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 4]
```diff
@@ -44,7 +49,7 @@ class AmpOptimWrapper(OptimWrapper):
     def __init__(self, loss_scale='dynamic', **kwargs):
         assert digit_version(TORCH_VERSION) >= digit_version('1.6.0'), (
             '`torch.cuda.amp` is only available when pytorch version >= 1.6')
-        assert torch.cuda.is_available(), (
+        assert torch.cuda.is_available() or is_npu_available(), (
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:52]:
```suggestion
        assert is_cuda_available() or is_npu_available(), (
```

File: mmengine/optim/optimizer/builder.py [hunk 5]
```diff
@@ -53,6 +54,10 @@ def build_optim_wrapper(model: nn.Module,
     constructor_type = optim_wrapper_cfg.pop('constructor',
                                              'DefaultOptimWrapperConstructor')
     paramwise_cfg = optim_wrapper_cfg.pop('paramwise_cfg', None)
+
+    if is_npu_available():
+        optim_wrapper_cfg['type'] = 'AmpOptimWrapper'
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/builder.py:59]: Need to add comments to explain the modification.

File: mmengine/dist/dist.py [hunk 6]
```diff
@@ -504,7 +508,8 @@ def broadcast_object_list(data: List[Any],
         if group is None:
             group = get_default_group()
 
-        if digit_version(TORCH_VERSION) >= digit_version('1.8.0'):
+        if digit_version(TORCH_VERSION) >= digit_version(
+                '1.8.0') and not is_npu_available():
```
Comment (zhouzaida) [inline mmengine/dist/dist.py:512]: If NPU does not support `broadcast_object_list`, adding an AssertionError may be proper.

File: mmengine/runner/amp.py [hunk 7]
```diff
@@ -5,7 +5,7 @@
 
 import torch
 
-from mmengine.device import get_device
+from mmengine.device import get_device, is_npu_available
```
Comment (zhouzaida) [inline mmengine/runner/amp.py:8]:
```suggestion
from mmengine.device import get_device, is_npu_available, is_cuda_available
```

File: mmengine/runner/amp.py [hunk 8]
```diff
@@ -86,7 +86,10 @@ def autocast(device_type: Optional[str] = None,
                 logger='current',
                 level=logging.WARNING)
 
-        if torch.cuda.is_available():
+        if is_npu_available():
+            with torch.npu.amp.autocast(enabled=enabled):
+                yield
+        elif torch.cuda.is_available():
```
Comment (zhouzaida) [inline mmengine/runner/amp.py:92]:
```suggestion
        elif is_cuda_available():
```

**Comment by zhouzaida** [comment]
Need to test hccl backend for dist module. https://github.com/open-mmlab/mmengine/tree/main/tests/test_dist

File: mmengine/dist/dist.py (hunk 6 above)
Comment (wangjiangben-hw) [reply mmengine/dist/dist.py:513]: using _broadcast_object_list instead

**Comment by wangjiangben-hw** [comment]
> * Test hccl backend for dist module. https://github.com/open-mmlab/mmengine/tree/main/tests/test_dist
> * Add `npu` method in BaseModel
> 
> https://github.com/open-mmlab/mmengine/blob/13484aae9fbe66dd98f6a645267d9965239b7ce8/mmengine/model/base_model/base_model.py#L198-L211

added  * Add `npu` method in BaseModel

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 9]
```diff
@@ -44,7 +49,7 @@ class AmpOptimWrapper(OptimWrapper):
     def __init__(self, loss_scale='dynamic', **kwargs):
         assert digit_version(TORCH_VERSION) >= digit_version('1.6.0'), (
             '`torch.cuda.amp` is only available when pytorch version >= 1.6')
-        assert torch.cuda.is_available(), (
+        assert is_cuda_available() or is_npu_available(), (
             '``AmpOptimizerWrapper`` is only available training on gpu')
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:53]:
```suggestion
            '``AmpOptimizerWrapper`` is only available training on gpu or npu')
```

File: mmengine/optim/optimizer/amp_optimizer_wrapper.py [hunk 10]
```diff
@@ -3,13 +3,18 @@
 
 import torch
 import torch.nn as nn
-from torch.cuda.amp import GradScaler
 
+from mmengine.device import is_npu_available, is_cuda_available
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/amp_optimizer_wrapper.py:7]:
```suggestion
from mmengine.device import is_cuda_available, is_npu_available
```

File: mmengine/model/base_model/base_model.py [hunk 11]
```diff
@@ -210,6 +210,25 @@ def cuda(
         self._set_device(torch.device(device))
         return super().cuda(device)
 
+    def npu(
+        self,
+        device: Optional[Union[int, str, torch.device]] = None,
```
Comment (HAOCHENYE) [inline mmengine/model/base_model/base_model.py:215]:
```suggestion
        device: Union[int, str, torch.device, None] = None,
```

**Review by HAOCHENYE — COMMENTED** [review]
`is_npu_avalialble` should also be added to `docs/en/device.rst` and `docs/zh_cn/device.rst`

**Comment by wangjiangben-hw** [comment]
> `is_npu_avalialble` should also be added to `docs/en/device.rst` and `docs/zh_cn/device.rst`

added

**Review by zhouzaida — APPROVED** [review]

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**